File: tdd/examples/java-unit-test-demo/src/main/java/com/concordusa/unittestdemo/service/RemainingCapacity.java
package com.concordusa.unittestdemo.service;

public class RemainingCapacity {
    private final MutableInteger remainingCapacity;

    RemainingCapacity(final MutableInteger spaceCapacity){
        remainingCapacity = spaceCapacity;
    }

    public Integer getRemainingCapacity(){
        return remainingCapacity.getCurrentValue();
    }

    public boolean allowEntry(final Integer partySize){
        return remainingCapacity.getCurrentValue() >= partySize;
    }

    public void admitParty(final Integer partySize){
        remainingCapacity.subtract(partySize);
    }

    public void removeParty(final Integer partySize){
        remainingCapacity.add(partySize);
    }

    public boolean admitPartyIfAllowedEntry(final Integer partySize){
        if(!allowEntry(partySize)){
            return false;
        }
        admitParty(partySize);
        return true;
    }

    public void enforceCapacity_a(){
        if(remainingCapacity.getCurrentValue() >= 0){
            return;
        }

        remainingCapacity.subtract(remainingCapacity.getCurrentValue());
    }

    public void enforceCapacity_b(){
        if(remainingCapacity.getCurrentValue() < 0){
            remainingCapacity.add(-1 * remainingCapacity.getCurrentValue());
        }
    }
}

File: practices/tdd/examples/net-core-unit-test-demo/src/UnitTestDemo/RemainingCapacity.cs
namespace UnitTestDemo
{
    public class RemainingCapacity
    {
        private MutableInteger remainingCapacity;

        public RemainingCapacity(MutableInteger spaceCapacity)
        {
            remainingCapacity = spaceCapacity;
        }

        public int GetRemainingCapacity()
        {
            return remainingCapacity.GetCurrentValue();
        }

        public bool AllowEntry(int partySize)
        {
            return remainingCapacity.GetCurrentValue() >= partySize;
        }

        public void AdmitParty(int partySize)
        {
            remainingCapacity.Subtract(partySize);
        }

        public void RemoveParty(int partySize)
        {
            remainingCapacity.Add(partySize);
        }

        public bool AdmitPartyIfAllowedEntry(int partySize)
        {
            if(!AllowEntry(partySize))
            {
                return false;
            }
            AdmitParty(partySize);
            return true;
        }

        public void EnforceCapacity_A()
        {
            if(remainingCapacity.GetCurrentValue() >= 0)
            {
                return;
            }

            remainingCapacity.Subtract(remainingCapacity.GetCurrentValue());
        }

        public void EnforceCapacity_B()
        {
            if(remainingCapacity.GetCurrentValue() < 0)
            {
                remainingCapacity.Add(-1 * remainingCapacity.GetCurrentValue());
            }
        }
    }
}
